lib/spack/spack/test_state.py

```diff
@@ -0,0 +1,114 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from types import ModuleType
+
+import pickle
+import pydoc
+import io
+import sys
+import multiprocessing
+
+import spack.architecture
+import spack.config
+
+
+patches = None
+
+
+def append_patch(patch):
+    global patches
+    if not patches:
+        patches = list()
+    patches.append(patch)
+
+
+def serialize(obj):
+    serialized_obj = io.BytesIO()
+    pickle.dump(obj, serialized_obj)
+    serialized_obj.seek(0)
+    return serialized_obj
+
+
+class SpackTestProcess(object):
+    def __init__(self, fn):
+        self.fn = fn
+
+    def _restore_and_run(self, fn, test_state):
+        test_state.restore()
+        fn()
+
+    def create(self):
+        test_state = TransmitTestState()
+        return multiprocessing.Process(
+            target=self._restore_and_run,
+            args=(self.fn, test_state))
+
+
+class TransmitTestState(object):
```

Comment:
I think this should be refactored to be more general, and combined with the package serialization code.

Could we have a `class SpackContextSerializer`, and then different types of objects could be registered for serialization? The package will be one of those, and monkeypatch patches would be another, etc. I think that will make the code a little more stable and easier to maintian/extend.

Is there something I'm missing that would make that difficult/impossible?

Reply:
> Could we have a class SpackContextSerializer, and then different types of objects could be registered for serialization?

If I understand, this class would have a `register` method and I could add all of the objects of interest. The issue with that is that most of the objects passed to `TransmitTestState` are serialized in different ways:

* The monkeypatches from the test must be explicitly serialized because the test repositories must be available when they are unpickled (so first the `Repo` must be reconstructed, then the patches deserialized).
* The `Repo` object is serialized without `pickle`: I just store the directories. In general if an object is not pickle-able but has a straightforward serialization method, I went with that.

(`config` and `architecture.platform` can be stored as attributes on the object: they serialize/deserialize without problems and order doesn't matter for them)

The second point isn't critical - I think `Repo` could be made pickle-able (without that each object would have to be registered along with a serialization method) - but the first means that a generic `register` method would also need a priority. In short, the generality required to support the specific serialization use cases for the Spack installation would make it fairly complex, and I don't know of other cases where this would be useful (e.g. typically you don't need a special object to serialize/deserialize: you can just pass objects as additional args to `multiprocessing.Process`).

Reply:
If the registered objects were held in a list, and deserialized in order, would that be sufficient?

What I really don't like is that the `PackageTransmitter` is responsible for passing the test state in addition to the package itself. That seems like it breaks separation of concerns.